Initial checkin of CPython fuzz tests.

ssbr opened the pull request:
This depends on changes that haven't landed in CPython yet. You can test it by replacing the git clone with:

    RUN git clone -b fix-issue-29505 --depth 1 https://github.com/ssbr/cpython.git cpython

The corresponding pull request to CPython is here: https://github.com/python/cpython/pull/2878

Those are the first fuzz tests I've ever written, and maybe are too trivial or something. Will get to better fuzz tests soon.

Extra notes:
  Google bug: b/37562550 (and there are others for other bits of CPython)
  Python bug: https://bugs.python.org/issue29505

kcc commented on projects/cpython3/build.sh:
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash -eu
+# Copyright 2016 Google Inc.
```
2017

kcc commented on projects/cpython3/build.sh:
```diff
@@ -0,0 +1,41 @@
+################################################################################
+
+# Workaround for distutils, which doesn't copy $CC except on OSX.
+export LDSHARED="clang -shared"
```
$CC instead of clang

kcc commented on projects/cpython3/gen_fuzz.py:
```diff
@@ -0,0 +1,47 @@
+"""Generate a fuzz test.
```
shouldn't this file be part of the python trunk?

kcc replied:
Also, do you really need to write this in python? 
Even a simple C include will work

ssbr replied:
> shouldn't this file be part of the python trunk?

I expect the code in this script to eventually be shared across both cpython2/cpython3, and probably also non-cpython fuzz tests that use the python runtime (e.g. simplejson). It could go in cpython, it just seems like there'll be slightly more duplication if I do. (Also, it'd be "dead code" in cpython's repo as it would never be executed).

Does that make sense?

> Also, do you really need to write this in python?
> Even a simple C include will work

I guess I am terrified of `#include SOME_CONSTANT`. Does that work if I do `-D 'SOME_CONSTANT="path/to/fuzz_foo.inc"'`? Or are you thinking I should use `SOME_CONSTANT` and pass `-D "source=$(cat path/to/fuzz_foo.inc)"`?

kcc replied:
> Also, it'd be "dead code" in cpython's repo as it would never be executed

That's the point. We encourage everyone to *not* have fuzz targets as dead code, and instead use them as part of CI
(w/o fuzzing): see https://github.com/google/oss-fuzz/blob/master/docs/ideal_integration.md

> I guess I am terrified of #include SOME_CONSTANT

Yea, don't do that. 

Also, why do you have this two-level complexity where you first define e.g. `fuzz_builtin_unicode`
and then call it in LLVMFuzzerTestOneInput. 

The way we propose is to just have LLVMFuzzerTestOneInput in your code for every API you want to test, 
see e.g. https://github.com/google/boringssl/tree/master/fuzz
Then you won't need this python script, and it will be very easy to incorporate the fuzz targets into the upstream build system
(e.g. using  https://github.com/llvm-mirror/llvm/tree/master/lib/Fuzzer/standalone or similar)

ssbr replied:
> The way we propose is to just have LLVMFuzzerTestOneInput in your code for every API you want to test,
see e.g. https://github.com/google/boringssl/tree/master/fuzz
Then you won't need this python script, and it will be very easy to incorporate the fuzz targets into the upstream build system
(e.g. using https://github.com/llvm-mirror/llvm/tree/master/lib/Fuzzer/standalone or similar)

The upstream build system runs all tests in a single process against a single binary, so this would be an ODR violation. That's the original reason for the separation.

But thinking on it more, I think I can move this upstream and share code without ODR violations, by making the function name parameterized, or by only invoking it once. Not quite sure how to invoke it (either as a script or a .h) from Python's build system though...

I'll think on this and try to come up with something that moves this to cpython.

kcc commented on projects/cpython3/gen_fuzz.py:
```diff
@@ -0,0 +1,47 @@
+    PyErr_Print();
+    abort();
+  }
+  return rv;
```
LLVMFuzzerTestOneInput should return 0 (other values are reserved for future)

kcc replied:
But I see that functions from https://github.com/ssbr/cpython/tree/fix-issue-29505/Modules/_fuzz only return 0, i.e. this is fine

ssbr commented:
I don't know how code review on github works, but I uploaded a second commit, PTAL.

ssbr commented:
Updated again, now moved the definition to CPython. It's a little wonky, so feel free to request more changes in the CPython branch. (I know the CPython devs will...)

Thanks for all your review so far, PTAL!

kcc commented on projects/cpython3/build.sh:
```diff
@@ -0,0 +1,42 @@
+
+# Build python to run the app.
+pushd cpython
+./configure --prefix=$OUT/ --with-ensurepip=no
```
Hm. Will this copy everything into $OUT? 
$OUT needs to contain only the fuzzer binaries and their dependencies, nothing extra

oliverchang replied:
On our fuzzing infrastructure, $OUY is different from the one that the binaries are built in. Will this still work, or are there hardcoded paths that will break?

ssbr replied:
I don't understand the question. It works if I run this command, for example:

    python infra/helper.py build_image cpython3 && python infra/helper.py build_fuzzers --sanitizer=address cpython3 && python infra/helper.py run_fuzzer cpython3 fuzz_builtin_unicode

oliverchang replied:
When we build, $OUT is something like /out

When we run this on our fuzzing infrastructure, $OUT is something completely different. Are there paths etc that depend on things being in /out ?

ssbr replied:
I don't know Python's built process well enough to answer either way. Is there a way to test this locally?

alex commented:
@ssbr Just reviewed your PR for CPython. Do you have a problem with me adding myself to the CCs for this? I'm a Python core developer and a member of the python security team.

ssbr commented on projects/cpython3/build.sh:
```diff
@@ -0,0 +1,42 @@
+
+# Build python to run the app.
+pushd cpython
+./configure --prefix=$OUT/ --with-ensurepip=no
```
CPython is dynamically linked, so we need it -- it isn't extra.

It's possible we can prune a lot of CPython though. (e.g. we don't need modules like TKinter/tk.)  Most of those should already be excluded from the build because they're optional and we probably didn't install their dependencies.

kcc replied:
Yea, try to not put anything other then the fuzzer binaries and their deps into $OUT -- otherwise ClusterFuzz will get confused and will try to execute something that it shouldn't. 

Is the python part of the change submitted?

ssbr replied:
The Python part isn't submitted -- the fuzz test needs to be written in C instead of C++, both to get approved and to integrate into the build system (which, annoyingly, passes -std=c99 even for C++ code). I tagged you in the review thread.

ssbr commented:
@alex (I can't figure out how to reply to your comment...), yes, that sounds like a great idea!

ssbr commented:
Those initial fuzz tests were merged into upstream, want to take another look?

kcc commented:
How much stuff will it put into $OUT? 
Only the stuff that is needed to run the fuzzers, or all of the python?

ssbr commented:
It will put all of Python into $OUT, which makes sense, because the fuzzer needs the entirety of the Python runtime to run. We could probably filter out the pure-python part of the stdlib if we put some effort into it.

kcc commented:
@oliverchang will this work from OSS-Fuzz POV if $OUT has the entirety of Python?

ssbr commented:
If not, I think we'll need to use the internal fuzz testing tech with bazel, because internally we can use Hermetic Python to avoid all this dynamic linking shenanigans. And then _all_ fuzz tests would need to be run on the internal infrastructure using the versions that are vendored in third_party. Maybe not optimal, but at least we'll be able to get them running.

Does that sound right?

(I just want to be sure I'm not wasting my time here...)

alex commented:
The current fuzzers we have only rely on libpython, so if we could get libpython as a `.a`, I guess that might work.

The desired fuzzers for things like json or cvs would be more of a challenge. Those modules are generally dynamically loaded `.so`s -- I don't think there's an existing way to link those in at compile time. (Though perhaps there is, `sys` and other builtin modules happen _somehow_)

oliverchang commented:
> @oliverchang will this work from OSS-Fuzz POV if $OUT has the entirety of Python?

Yep. We look for ELF files with LLVMFuzzerTestOneInput in its contents so this should work, 

**But**, as for https://github.com/google/oss-fuzz/pull/731#discussion_r137872849, 

I built this locally, and did

```
python infra/helper.py build_image cpython3
python infra/helper.py build_fuzzers cpython3
python infra/helper.py shell cpython3

$ mkdir /out2
$ mv /out/* /out2
$ /out2/fuzz_builtin_float
```

And see:

```
Fatal Python error: Py_Initialize: Unable to get the locale encoding
ModuleNotFoundError: No module named 'encodings'

Current thread 0x00007f6c4dc9e780 (most recent call first):
==62== ERROR: libFuzzer: deadly signal
    #0 0x533243 in __sanitizer_print_stack_trace /src/llvm/projects/compiler-rt/lib/asan/asan_stack.cc:38
    #1 0x57eeb5 in fuzzer::Fuzzer::CrashCallback() /src/libfuzzer/FuzzerLoop.cpp:196:5
    #2 0x57ee4c in fuzzer::Fuzzer::StaticCrashSignalCallback() /src/libfuzzer/FuzzerLoop.cpp:175:6
   ....
```

So it does look like the build expects things to always exist in the same `$OUT` that's passed to configure. @ssbr do you see any way around this so that we can relocate $OUT to anywhere and things will still work?

(We probably should improve our `run_fuzzer` command to always to this to catch more errors like this).

ssbr commented:
> (We probably should improve our run_fuzzer command to always to this to catch more errors like this).
Filed https://github.com/google/oss-fuzz/issues/837

> @ssbr do you see any way around this so that we can relocate $OUT to anywhere and things will still work?

Is there an environment variable we can use to know the true later value of $OUT?

I can think of some hacky ways to try to fix this, but I suspect this is fighting a losing battle. It'd be better to just use the same $OUT or something, or else to give up and use a different fuzz runner (like one where we can use hermetic python / bazel).

kcc commented:
You can use bazel here (grpc does this: https://github.com/google/oss-fuzz/blob/master/projects/grpc/build.sh)

oliverchang commented:
> Is there an environment variable we can use to know the true later value of $OUT?

Unfortunately not. This can change at any time on our actual fuzzing infrastructure.

ssbr commented:
> You can use bazel here (grpc does this: https://github.com/google/oss-fuzz/blob/master/projects/grpc/build.sh)

We can't use hermetic python though, right? (Requires a patched glibc or something.)

kcc commented:
That I don't know :(

ssbr commented:
So here is my suggestion: I've asked the python-dev IRC channel if they have any ideas on how to make it work with how oss-fuzz builds/tests binaries. If it can be made to work, great! If not, I propose we don't use oss-fuzz to fuzz this project, and instead fuzz the version stored in google's version control system with our internal fuzzing test tools.

kcc commented:
That's up to you, of course. 
One of the benefits of oss-fuzz is that anyone on the python team (not only googlers) can see the results.

oliverchang commented:
> Is there an environment variable we can use to know the true later value of $OUT?

> I can think of some hacky ways to try to fix this (PYTHONPATH environment variable, for starters), but I > suspect this is fighting a losing battle. It'd be better to just use the same $OUT or something, or else to > give up and use a different fuzz runner (like one where we can use hermetic python / bazel).

@ssbr Ah, I think I may have misunderstood your comment.

If you mean an environment variable during runtime, we can certainly provide one. It does look like PYTHONPATH fixes this.

Following my previous example,

```
PYTHONPATH=/out2/lib/python3.7 ./fuzz_builtin_float
```

I guess the fuzz target can initialize this path during initialization?

ssbr commented:
You had me right the first time -- I don't trust PYTHONPATH as a solution.

Moreover I don't see a meaningful difference between running the test in oss-fuzz vs in google's internal things. Either way the bug reports will make it upstream and we will hopefully catch security issues. But if we run it in some internal not-open-source tests, then we have the benefit of it working with the usual google build system which has already applied every hack in the book to make python hermetic and copyable, so we can piggyback on that work and stop worrying.

ssbr commented:
I'm probably going to talk it over with @gpshead and/or Thomas Wouters (unsure of github nick), who work on Python at Google.

oliverchang commented:
> Moreover I don't see a meaningful difference between running the test in oss-fuzz vs in google's internal things. 

I'm not too familiar with how things are internally these days, but OSS-Fuzz will likely reduce the amount of work you have to do wrt bug reports making their way upstream. We continuously test on the latest upstream revisions, and automatically handle triage, de-duping and reporting.

ssbr commented:
My assumption was that ISE-team, or whoever else runs the fuzz tests, will handle it. Much like oss-fuzz. 😅

gpshead commented:
fwiw - I object to us running any of this internally at Google. We need to be part of the main oss-fuzz project pulling from upstream revisions. Doing this testing within our blackhole of internal stuff adds more work for us internally (read: which we're not going to do) and wouldn't provide results feedback to the upstream CPython project in a useful timely manner.

We must figure out how to get this to build and run on the external oss-fuzz infrastructure.

Dor1s commented:
Ping. What's the status here?

ssbr commented:
I wasn't able to figure out something that would work except possibly `PYTHONPATH`, which I never tried. I also burned out and left the team (youtube-security), so if nobody else picks this up it's dead.

Dor1s commented:
@ssbr, thanks for the info!

@gpshead, are you interested in getting this landed?

Dor1s commented:
@durin42 @markus-kusano @gpshead ping

Lucas-C commented:
I took a look at this, and based on @markus-kusano PR in #2031 and with his help, I think I managed to get it working.
My branch is here: https://github.com/Lucas-C/oss-fuzz/commits/cpython3

The following commands work fine now:
```
python infra/helper.py run_fuzzer --sanitizer address cpython3 fuzz_builtin_float
python infra/helper.py run_fuzzer --sanitizer address cpython3 fuzz_builtin_int
python infra/helper.py run_fuzzer --sanitizer address cpython3 fuzz_builtin_unicode
```

At some point I got the `ModuleNotFoundError: No module named 'encodings'` error, and it was due to the shared modules (Python builtin `.py` libs) not being installed.

Should I do more tests and can I submit a PR ?

markus-kusano commented:
ping @Dor1s 

Were you able to move the fuzz target out of /out and into a new directory and still have it work? see https://github.com/google/oss-fuzz/pull/731#issuecomment-328202107. If so: amazing :)

If that is working then I can think of anything else that is blocking this, but someone from OSS-Fuzz would have to take a look. Probably easiest to just open a new pull request for your OSS-Fuzz changes.

Dor1s commented:
No, I didn't try that (and don't think that I ever said that I was going to :) )

markus-kusano commented:
Sorry for the lack of context in that ping :) Just was wondering if someone from OSS-Fuzz could see if this was working, though, I'm not sure if all the CPython changes are in the forked repo.

Dor1s commented:
> libfuzzer generates input like 52**52**32 after a few hours, and they trigger a timeout. As this valid Python code, is it possible to ignore those timeout in some way ?

I'm afraid you'll have to blacklist things like this in a fuzz target OR patch the Python code using `#ifdef FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION`, see http://llvm.org/docs/LibFuzzer.html#fuzzer-friendly-build-mode

ethanhs commented:
@Lucas-C any update on that? I'm excited to see CPython fuzzed! Thank you for working on it :)

jonathanmetzman commented:
Superseeded by #2493
